Run basic News Article tests against GraphQL, too

mike3985 opened the pull request:
## What
1. Create a test fixture based on the vanilla News Article content schema example.
2. Run the basic News Article system tests against the fixture.

## Why

We're working towards having GraphQL producing content as close to Content Store's as is possible and reasonable. A bit more test coverage is going to help.

[Trello card](https://trello.com/c/HZRt4r50/1666-make-meta-tags-etc-of-graphql-version-of-news-articles-the-same-as-regular-version), [Jira issue PP-2906](https://gov-uk.atlassian.net/browse/PP-2906)

mike3985 committed:
Rename copy-pastad describe

mike3985 committed:
GraphQL version of news_article schema example

This is a take on the "news_article" example content schema that some
Frontend specs make use of. We've made a GraphQL equivalent to allow us
to run the same tests with both responses.

This GraphQL-flavoured example only includes fields that are currently
requested by our `Graphql::EditionQuery`. That does imply a maintainance
burdern in keeping the query and the JSON fixtures in sync, but we
haven't figured out a smarter approach to that yet.

In the longer-term we might want these to live somewhere central like
the content schema examples. For now, there's no infrastructure for that
and these news articles examples are only relevant to Frontend as
a client.

mike3985 committed:
Move fetch_graphql_fixture to dedicated helper

mike3985 committed:
`graphql_has_example_item` test helper

This follows a similar pattern to the existing helper
`content_store_has_example_item`... in that it:

* gets its content item data from a static JSON example file,
* stubs the query that Frontend will make to the backend (in this case
  that's Publishing API instead of Content Store), and
* returns the parsed "content item" hash to the caller

mike3985 committed:
Run basic News Article tests against GraphQL, too

yndajas commented on spec/fixtures/graphql/news_article_christmas_2016.json:
```diff
@@ -0,0 +1,145 @@
+      "description": "Theresa May sends her best wishes to everyone celebrating Christmas in the UK and around the world.",
+      "details": {
+        "body": "<div class=\"govspeak\"><p>Prime Minister Theresa May said:</p><blockquote>  <p>This year, the United Kingdom has had much to celebrate. Her Majesty The Queen celebrated her 90th birthday, surrounded by the Royal Family and well-wishers from across our four nations, the Commonwealth, and the world.</p>  <p>Four years after the success of London 2012, our Olympic and Paralympic at ...
+        "change_history": [
+           {
+              "note": "Added translation",
+              "public_timestamp": "2016-12-25T00:15:02.000+00:00"
+           }
+        ],
```
**question:** did you get the data from the live version rather than the content schemas example? Some of this is missing from the latter

mike3985 replied:
I started with the GraphQL response in integration and then updated some values where they differed enough to upset the tests

mike3985 replied:
updated some values -> copied some values from the content schema example

yndajas commented on spec/system/news_article_spec.rb:
```diff
@@ -54,4 +54,18 @@
+
+  context "when content item is from Publishing API's GraphQL" do
+    let(:content_item) { graphql_has_example_item("news_article_christmas_2016") }
+    let(:path) { "#{content_item.fetch('base_path')}?graphql=true" }
```
**question:** is there a reason for only hard coding it here (unlike the Content Store example)?

mike3985 replied:
for _not_ hard coding it here? that's the result of a chaotic development process

mike3985 replied:
they're now both hard coded

leenagupte commented on spec/requests/news_article_spec.rb:
```diff
@@ -29,11 +28,8 @@
-
       before do
-        stub_publishing_api_graphql_content_item(Graphql::EditionQuery.new(base_path).query, graphql_fixture)
+        base_path = graphql_has_example_item("news_article").fetch("base_path")
```
This should be in a `let` block. At the moment I think the test is picking up the value of `base_path` from line 9. If the test is passing, I wonder it's actually testing anything from the fixture?

mike3985 replied:
I've made an attempt at making this better to read. None of the variables from the `let`s were actually being used in the example blocks, so I've gone ahead and moved all of those setup variables into their respective `before`s

leenagupte commented on spec/system/news_article_spec.rb:
```diff
@@ -3,8 +3,8 @@
 
-  context "when visiting a page" do
-    before { visit "/government/news/christmas-2016-prime-ministers-message" }
+  shared_examples "a news article page" do
```
I think these examples should be moved to the bottom of the test file as it's a little jarring to have the history mode tests between the shared example definitions and the tests that run them.

mike3985 replied:
I've taken your point but resolved it the opposite way: I've moved the contexts that actually use the shared examples above the history mode contexts (based on the understanding that the more basic/generic a test, the nearer the top of the test file it should be)

leenagupte reviewed (commented):
This looks good to me. I have a couple of minor inline comments.

mike3985 committed:
Local variables for test setup instead of `let`s

To me, this doesn't feel like a bad idea in general. `let`s are useful
for defining variables used in example blocks, but are probably clutter
in this case.

Right now there's a pleasant consistency between the Content Store and
GraphQL examples in this spec, but I'm en route to changing how these
values are derived in the GraphQL context and I think moving the
variables to where they're used is going to help maintain the
skim-readability.

leenagupte commented on spec/system/news_article_spec.rb:
```diff
@@ -1,4 +1,4 @@
-RSpec.describe "Fatality Notice" do
+RSpec.describe "News Article" do
```
Thanks for fixing this. I've just realised that there is a typo in the commit message 🍝

mike3985 replied:
It's quite rare for my commit messages to not have typos, so you're probably right only I can't see it. But the pasta part was intentional, https://en.wikipedia.org/wiki/Copypasta#Technology 😋

leenagupte replied:
In the future I wouldn't put intentional typos in as it just looks odd.

mike3985 replied:
Ha yeah, it wasn't the typo that was intentional, just the reference to copy-pastaing. I suppose I made some poor punctuation choices that resulted in what looked like a typo!

leenagupte replied:
Next time please fix the typos (intentional or not) before merging.

leenagupte reviewed (approved):
👍 

There's a typo in the commit message that should be fixed, but doesn't need a re-review.
